contracts/SwapAndBridge.sol

```diff
@@ -291,4 +310,178 @@ contract UniversalSwapAndBridge is SwapAndBridgeBase {
+
+        _swapToken.safeTransferFrom(msg.sender, address(this), swapTokenAmount);
+        _swapAndBridge(
+            routerCalldata,
+            swapTokenAmount,
+            minExpectedInputTokenAmount,
+            depositData,
+            _swapToken,
+            acrossInputToken
+        );
+    }
+
+    /**
+     * @notice Swaps an EIP-3009 token on this chain via specified router before submitting Across deposit atomically.
+     * Caller can specify their slippage tolerance for the swap and Across deposit params.
+     * @dev If swapToken does not implement `receiveWithAuthorization` to the specifications of EIP-3009, this call will revert.
+     * @param swapToken Address of the token that will be swapped for acrossInputToken.
+     * @param acrossInputToken Address of the token that will be bridged via Across as the inputToken.
+     * @param routerCalldata ABI encoded function data to call on router. Should form a swap of swapToken for
+     * enough of acrossInputToken, otherwise this function will revert.
+     * @param swapTokenAmount Amount of swapToken to swap for a minimum amount of depositData.inputToken.
+     * @param minExpectedInputTokenAmount Minimum amount of received depositData.inputToken that we'll submit bridge
+     * deposit with.
+     * @param depositData Specifies the Across deposit params we'll send after the swap.
+     * @param validAfter The unix time after which the `receiveWithAuthorization` signature is valid.
+     * @param validBefore The unix time before which the `receiveWithAuthorization` signature is valid.
+     * @param nonce Unique nonce used in the `receiveWithAuthorization` signature.
+     * @param v v of the EIP-3009 signature.
+     * @param r r of the EIP-3009 signature.
+     * @param s s of the EIP-3009 signature.
+     */
+    function swapAndBridgeWithAuthorization(
+        IERC20Auth swapToken,
+        IERC20 acrossInputToken,
+        bytes calldata routerCalldata,
+        uint256 swapTokenAmount,
+        uint256 minExpectedInputTokenAmount,
+        DepositData calldata depositData,
+        uint256 validAfter,
+        uint256 validBefore,
+        bytes32 nonce,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external nonReentrant {
+        // While any contract can vacuously implement `transferWithAuthorization` (or just have a fallback),
+        // If tokens were not sent to this contract, the call to `transferFrom` in _swapAndBridge will revert.
+        swapToken.receiveWithAuthorization(
+            msg.sender,
+            address(this),
+            swapTokenAmount,
+            validAfter,
+            validBefore,
+            nonce,
+            v,
+            r,
+            s
+        );
+        IERC20 _swapToken = IERC20(address(swapToken)); // Cast IERC20Auth to IERC20
+
+        uint256 tokenBalance = _swapToken.balanceOf(address(this));
+        if (tokenBalance < swapTokenAmount)
```

Comment:
Do you think we need this check? Won't `_swapAndBridge` fail if we have insufficient balance? Or do you think we want to just catch it earlier?

I guess i'm thinking, what's the difference between checking the balance here after calling `receiveWithAuthorization()` but not checking the balance after calling `safeTransferFrom` in the other two methods? Is it because in the latter we are more familiar with that function and knowing it **should just work**, whereas `receiveWithAuthorization` is a newer function that we're not used to?

Reply:
Yeah... That's a good point. I added this since I have the expectation that `receiveWithAuthorization` is just an uncommon function in most ERC20s, but whether a token implements this function can probably just be checked off-chain, so the check is just unnecessary.